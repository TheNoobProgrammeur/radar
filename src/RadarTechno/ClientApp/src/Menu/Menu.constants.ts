export const HOME = 'Home';
export const WORKFLOW = 'Workflow';
export const TECHNOLOGIES = 'Technologies';
export const ALL_TECHNOLOGIES = 'All Technologies';
export const USERS = 'Users';
export const ENTITIES = 'Entities';

export const menuPositions = {
  [HOME]: 0,
  [WORKFLOW]: 1,
  [TECHNOLOGIES]: 2,
  [ALL_TECHNOLOGIES]: 3,
  [USERS]: 4,
  [ENTITIES]: 5
};
