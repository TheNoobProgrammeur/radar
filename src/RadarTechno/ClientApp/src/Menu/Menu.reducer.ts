import { SET_POSITION, SET_VISIBLE } from '../constants/actions';

interface IState {
  position: number;
  visible: boolean;
}

interface IAction {
  type: string;
  position?: number;
  visible?: boolean;
}

export const initState = {
  position: 0,
  visible: false,
};

export const menuReducer = (state: IState, action: IAction) => {
  switch (action.type) {
    case SET_VISIBLE:
      return { ...state, visible: action.visible };
    case SET_POSITION:
      return { ...state, position: action.position, visible: false };
    default:
      return state;
  }
};
