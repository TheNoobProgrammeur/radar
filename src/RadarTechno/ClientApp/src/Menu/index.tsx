export { Menu } from './Menu';
export { EnhancedMenu } from './Menu.container';
