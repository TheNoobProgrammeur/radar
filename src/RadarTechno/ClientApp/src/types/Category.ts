export enum Category {
  frameworks = 'Languages and Frameworks',
  patterns = 'Patterns',
  platforms = 'Platforms',
  tools = 'Tools',
  infrastructures = 'Infrastructures',
  practices = 'Practices'
}
