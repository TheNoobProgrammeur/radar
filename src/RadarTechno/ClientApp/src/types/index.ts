export * from './Category';
export * from './CurrentUser';
export * from './FormFields';
export * from './LoginFields';
export * from './Paging';
export * from './Status';
