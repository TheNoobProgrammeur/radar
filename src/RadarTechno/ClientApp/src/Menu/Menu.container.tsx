import React, { useContext, useEffect } from 'react';
import useEnhancedReducer from 'react-enhanced-reducer-hook';
import thunkMiddleware from 'redux-thunk';
import logMiddleware from '../logMiddleware';
import { UserContext } from '../User';
import { Menu } from './Menu';
import { handleClick, initPosition, setPosition } from './Menu.action';
import { initState, menuReducer } from './Menu.reducer';

export const EnhancedMenu = () => {
  const { currentUser } = useContext(UserContext);
  const [state, dispatch] = useEnhancedReducer(menuReducer, initState, [
    thunkMiddleware,
    logMiddleware,
  ]);

  useEffect(() => {
    dispatch(initPosition());
  }, []);

  return (
    <Menu
      hideMenuButton={currentUser == null}
      handleClick={() => dispatch(handleClick())}
      user={currentUser}
      isMenuVisible={state.visible}
      position={state.position}
      updatePosition={(position: number) => dispatch(setPosition(position))}
    />
  );
};
