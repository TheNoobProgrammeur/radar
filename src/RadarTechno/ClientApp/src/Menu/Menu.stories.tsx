import React from 'react';

import { action } from '@storybook/addon-actions';
import { boolean, number, object, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import StoryRouter from 'storybook-react-router';

import { Menu } from '../Menu';

const fakeEntities = [{
  id:'5a4e2821a111b016c4cc5804',
  name:'AXA France',
  adminList:['Test@axa.fr'],
  technologies: 1,
  version:0,
}];

const fakeUser = {
  id:'1',
  email:'Test',
  role: 'user',
  entity: fakeEntities[0],
  entityList: ['5a4e2821a111b016c4cc5804'],
  token:'token',
};

storiesOf('Menu', module)
  .addDecorator(withKnobs)
  .addDecorator(StoryRouter())
  .add('Menu', () => <Menu 
    hideMenuButton={boolean('hideMenuButton', false)}
    user={object('user', fakeUser)}
    handleClick={action('handleClick')}
    position={number('position', 0)}
    isMenuVisible={boolean('isMenuVisible', true)}
    updatePosition={action('updatePosition')}
  />
);
