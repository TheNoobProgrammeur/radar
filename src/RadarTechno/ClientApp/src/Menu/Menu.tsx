import { NavBar, NavBarItem, Title } from '@axa-fr/react-toolkit-all';
import React from 'react';
import { Link } from 'react-router-dom';
import { ICurrentUser } from '../types';
import { ALL_TECHNOLOGIES, ENTITIES, HOME, menuPositions, TECHNOLOGIES, USERS, WORKFLOW } from './Menu.constants';

import './Menu.scss';

export const Menu = ({
  user,
  handleClick,
  isMenuVisible,
  position,
  updatePosition,
  hideMenuButton
}: {
  user: ICurrentUser,
  handleClick: (...args: any[]) => any,
  isMenuVisible: boolean,
  position: number,
  updatePosition: (position: number) => any,
  hideMenuButton?: boolean
}) => (
  <>
    {user && <NavBar positionInit={position} isVisible={isMenuVisible} onClick={handleClick}>
      <NavBarItem
        key="home"
        label={HOME}
        actionElt={
          <Link onClick={() => updatePosition(menuPositions[HOME])} className="af-nav__link" to="/">
            {HOME}
          </Link>
        }
      />
      <NavBarItem
        key="workflow"
        label={WORKFLOW}
        actionElt={
          user.entity.workflowUrl ? 
          <a className="af-nav__link" href={user.entity.workflowUrl} rel="noopener" target="_blank">{WORKFLOW}</a> :
          <Link onClick={() => updatePosition(menuPositions[WORKFLOW])} className="af-nav__link" to="/workflow">
            {WORKFLOW}
          </Link>
        }
      />
      <NavBarItem
        key="technologies"
        label={TECHNOLOGIES}
        actionElt={
          <Link onClick={() => updatePosition(menuPositions[TECHNOLOGIES])} className="af-nav__link" to="/technologies">
            Technologies in {user.entity.name}
          </Link>
        }
      />
      {(user.role === 'admin' || user.role === 'root') && <NavBarItem
        key="all-technologies"
        label={ALL_TECHNOLOGIES}
        actionElt={
          <Link onClick={() => updatePosition(menuPositions[ALL_TECHNOLOGIES])} className="af-nav__link" to="/all-technologies">
            List of technologies
          </Link>
        }
      />}
      {(user.role === 'admin' || user.role === 'root') && <NavBarItem
        key="users"
        label={USERS}
        actionElt={
          <Link onClick={() => updatePosition(menuPositions[USERS])} className="af-nav__link" to="/users">
            {USERS}
          </Link>
        }
      />}
      {user.role === 'root' && <NavBarItem
        key="entities"
        label={ENTITIES}
        actionElt={
          <Link onClick={() => updatePosition(menuPositions[ENTITIES])} className="af-nav__link" to="/entities">
            {ENTITIES}
          </Link>
        }
      />}
    </NavBar>}
    <Title
      title="Technology Radar"
      subtitle={`${user ? user.entity.name : ''}`}
      toggleMenu={handleClick}
      classModifier={hideMenuButton ? 'hide-menu-button' : 'show-menu-button'} 
    />
  </>
);
