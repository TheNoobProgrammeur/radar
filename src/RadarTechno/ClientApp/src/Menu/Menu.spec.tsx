import { NavBar } from '@axa-fr/react-toolkit-all';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { ICurrentUser } from '../types';
import { Menu } from './Menu';

configure({ adapter: new Adapter() });

describe('Menu component test', () => {

  test('Menu component without user matches snapshot', () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={['/']} keyLength={0}>
        <Menu user={null} isMenuVisible={true} handleClick={() => null}
          position={0} updatePosition={null} hideMenuButton={true}/>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Menu component with user matches snapshot', () => {
    const user: ICurrentUser = {
      email: 'email',
      entity: {
        id:'5a4e2821a111b016c4cc5804',
        name:'AXA France',
        adminList:['Test@axa.fr'],
        technologies: null,
        version:0,
        workflowUrl: 'url'
      },
      id: 'id',
      token: 'token',
      role: 'root'
    };
    const wrapper = shallow(
      <MemoryRouter initialEntries={['/']} keyLength={0}>
        <Menu user={user} isMenuVisible={true} handleClick={() => null} 
          position={0} updatePosition={null} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Menu component with user doesn\'t render NavBar', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']} keyLength={0}>
        <Menu user={null} isMenuVisible={true} handleClick={() => null} 
          position={0} updatePosition={null} hideMenuButton={true} />
      </MemoryRouter>
    );
    expect(wrapper.find(NavBar)).toHaveLength(0);
  });

  test('Menu component with user renders NavBar', () => {
    const user: ICurrentUser = {
      email: 'email',
      entity: {
        id:'5a4e2821a111b016c4cc5804',
        name:'AXA France',
        adminList:['Test@axa.fr'],
        technologies: null,
        version:0,
        workflowUrl: 'url'
      },
      id: 'id',
      token: 'token',
      role: 'root'
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']} keyLength={0}>
        <Menu user={user} isMenuVisible={true} handleClick={() => null} 
          position={0} updatePosition={null} />
      </MemoryRouter>
    );
    expect(wrapper.find(NavBar)).toHaveLength(1);
  });
});
