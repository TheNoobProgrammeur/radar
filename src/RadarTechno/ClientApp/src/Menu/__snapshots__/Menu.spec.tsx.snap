// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component test Menu component with user matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryRouter
    initialEntries={
      Array [
        "/",
      ]
    }
    keyLength={0}
  >
    <Menu
      handleClick={[Function]}
      isMenuVisible={true}
      position={0}
      updatePosition={null}
      user={
        Object {
          "email": "email",
          "entity": Object {
            "adminList": Array [
              "Test@axa.fr",
            ],
            "id": "5a4e2821a111b016c4cc5804",
            "name": "AXA France",
            "technologies": null,
            "version": 0,
            "workflowUrl": "url",
          },
          "id": "id",
          "role": "root",
          "token": "token",
        }
      }
    />
  </MemoryRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Menu
        handleClick={[Function]}
        isMenuVisible={true}
        position={0}
        updatePosition={null}
        user={
          Object {
            "email": "email",
            "entity": Object {
              "adminList": Array [
                "Test@axa.fr",
              ],
              "id": "5a4e2821a111b016c4cc5804",
              "name": "AXA France",
              "technologies": null,
              "version": 0,
              "workflowUrl": "url",
            },
            "id": "id",
            "role": "root",
            "token": "token",
          }
        }
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "handleClick": [Function],
        "isMenuVisible": true,
        "position": 0,
        "updatePosition": null,
        "user": Object {
          "email": "email",
          "entity": Object {
            "adminList": Array [
              "Test@axa.fr",
            ],
            "id": "5a4e2821a111b016c4cc5804",
            "name": "AXA France",
            "technologies": null,
            "version": 0,
            "workflowUrl": "url",
          },
          "id": "id",
          "role": "root",
          "token": "token",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Menu
          handleClick={[Function]}
          isMenuVisible={true}
          position={0}
          updatePosition={null}
          user={
            Object {
              "email": "email",
              "entity": Object {
                "adminList": Array [
                  "Test@axa.fr",
                ],
                "id": "5a4e2821a111b016c4cc5804",
                "name": "AXA France",
                "technologies": null,
                "version": 0,
                "workflowUrl": "url",
              },
              "id": "id",
              "role": "root",
              "token": "token",
            }
          }
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleClick": [Function],
          "isMenuVisible": true,
          "position": 0,
          "updatePosition": null,
          "user": Object {
            "email": "email",
            "entity": Object {
              "adminList": Array [
                "Test@axa.fr",
              ],
              "id": "5a4e2821a111b016c4cc5804",
              "name": "AXA France",
              "technologies": null,
              "version": 0,
              "workflowUrl": "url",
            },
            "id": "id",
            "role": "root",
            "token": "token",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Menu component test Menu component without user matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryRouter
    initialEntries={
      Array [
        "/",
      ]
    }
    keyLength={0}
  >
    <Menu
      handleClick={[Function]}
      hideMenuButton={true}
      isMenuVisible={true}
      position={0}
      updatePosition={null}
      user={null}
    />
  </MemoryRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Menu
        handleClick={[Function]}
        hideMenuButton={true}
        isMenuVisible={true}
        position={0}
        updatePosition={null}
        user={null}
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "handleClick": [Function],
        "hideMenuButton": true,
        "isMenuVisible": true,
        "position": 0,
        "updatePosition": null,
        "user": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Menu
          handleClick={[Function]}
          hideMenuButton={true}
          isMenuVisible={true}
          position={0}
          updatePosition={null}
          user={null}
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleClick": [Function],
          "hideMenuButton": true,
          "isMenuVisible": true,
          "position": 0,
          "updatePosition": null,
          "user": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
